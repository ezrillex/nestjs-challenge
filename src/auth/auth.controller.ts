import {
  Body,
  Controller,
  HttpCode,
  HttpStatus,
  Post,
  Request,
  UseGuards,
} from '@nestjs/common';
import { SignupUserDto } from './dto/signup-user.dto';
import { AuthService } from './auth.service';
import { LoginUserDto } from './dto/login-user.dto';
import { ForgotPasswordDto } from './dto/forgot-password.dto';
import { ResetPasswordDto } from './dto/reset-password.dto';
import { Public } from '../common/decorators/public.decorator';
import { Throttle, ThrottlerGuard } from '@nestjs/throttler';

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Public()
  @Post('signup')
  async signup(@Body() data: SignupUserDto) {
    const created_user = await this.authService.registerUser(data);
    return created_user;
  }

  @Public()
  @HttpCode(HttpStatus.OK)
  @Post('login')
  async login(@Body() data: LoginUserDto) {
    return this.authService.loginUser(data);
  }

  @HttpCode(HttpStatus.NO_CONTENT)
  @Post('logout')
  async logout(@Request() request: Request) {
    const userId = request['user'].id;
    return this.authService.logoutUser(userId);
  }

  @Public()
  @HttpCode(HttpStatus.OK)
  @Post('forgot')
  async forgot(@Body() data: ForgotPasswordDto) {
    return this.authService.forgotPassword(data);
  }

  @Public()
  @Throttle({ default: { limit: 3, ttl: 60000 } })
  @UseGuards(ThrottlerGuard)
  @HttpCode(HttpStatus.OK)
  @Post('reset')
  async reset(@Body() data: ResetPasswordDto) {
    return this.authService.changePassword(data);
  }
}
