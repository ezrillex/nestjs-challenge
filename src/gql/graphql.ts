
/*
 * -------------------------------------------------------
 * THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
 * -------------------------------------------------------
 */

/* tslint:disable */
/* eslint-disable */

export class GetProductsInput {
    first?: Nullable<number>;
    offset?: Nullable<number>;
    categoryFilter?: Nullable<string[]>;
    search?: Nullable<string>;
    likedOnly?: Nullable<boolean>;
    omitImages?: Nullable<boolean>;
    sort?: Nullable<string>;
}

export class CreateProductInput {
    name: string;
    description: string;
    categories?: Nullable<string[]>;
    variations: CreateVariationInput[];
}

export class CreateVariationInput {
    title: string;
    price: number;
    stock: number;
}

export class UpdateProductInput {
    id: string;
    name?: Nullable<string>;
    description?: Nullable<string>;
    categories?: Nullable<string[]>;
    is_published?: Nullable<boolean>;
}

export class UpdateProductVariationInput {
    id: string;
    title?: Nullable<string>;
    price?: Nullable<number>;
    stock?: Nullable<number>;
}

export class CreateProductVariationInput {
    product_id: string;
    title: string;
    price: number;
    stock: number;
}

export class CreateCategoryInput {
    name: string;
}

export class Images {
    __typename?: 'Images';
    id: string;
    url: string;
    created_at: string;
    product_variation_id?: Nullable<string>;
}

export class ProductVariations {
    __typename?: 'ProductVariations';
    id: string;
    title: string;
    price: number;
    stock: number;
    images?: Nullable<Images[]>;
}

export class Categories {
    __typename?: 'Categories';
    id: string;
    name: string;
    products: Products[];
}

export class Products {
    __typename?: 'Products';
    id: string;
    is_published: boolean;
    is_deleted: boolean;
    name: string;
    description: string;
    likes: ProductVariations[];
    variations: ProductVariations[];
    categories: Categories[];
    created_by: string;
    created_at: DateTime;
    last_updated_by?: Nullable<string>;
    last_updated_at?: Nullable<string>;
}

export class Users {
    __typename?: 'Users';
    id: string;
    first_name: string;
    last_name: string;
    email: string;
    likes_products: LikesOfProducts[];
}

export class LikesOfProducts {
    __typename?: 'LikesOfProducts';
    id: string;
    liked_by: Users;
    likes_product_variation: ProductVariations;
}

export class CartItems {
    __typename?: 'CartItems';
    id: string;
    cart_owner: Users;
    product_variation: ProductVariations;
    quantity: number;
}

export class OrderItems {
    __typename?: 'OrderItems';
    id: string;
    product_variation: ProductVariations;
    quantity: number;
    price_purchased_at: number;
}

export class Orders {
    __typename?: 'Orders';
    id: string;
    user: Users;
    paymentStatus: string;
    orderStatus: string;
    order_items: OrderItems[];
}

export abstract class IQuery {
    __typename?: 'IQuery';

    abstract getProducts(GetProductsInput: GetProductsInput): Nullable<Products[]> | Promise<Nullable<Products[]>>;

    abstract getProductById(id: string): Nullable<Products> | Promise<Nullable<Products>>;

    abstract getLikes(): Nullable<LikesOfProducts[]> | Promise<Nullable<LikesOfProducts[]>>;

    abstract getCartItems(): Nullable<CartItems[]> | Promise<Nullable<CartItems[]>>;

    abstract getOrders(): Nullable<Orders[]> | Promise<Nullable<Orders[]>>;

    abstract getOrder(order_id: string): Nullable<Orders> | Promise<Nullable<Orders>>;

    abstract getClientOrders(client_id?: Nullable<string>): Nullable<Orders[]> | Promise<Nullable<Orders[]>>;
}

export abstract class IMutation {
    __typename?: 'IMutation';

    abstract createProduct(CreateProductInput: CreateProductInput): Nullable<string> | Promise<Nullable<string>>;

    abstract updateProduct(UpdateProductInput: UpdateProductInput): Nullable<string> | Promise<Nullable<string>>;

    abstract updateProductVariation(UpdateProductVariationInput: UpdateProductVariationInput): Nullable<string> | Promise<Nullable<string>>;

    abstract createProductVariation(CreateProductVariationInput: CreateProductVariationInput): Nullable<string> | Promise<Nullable<string>>;

    abstract deleteProductVariation(variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract deleteProduct(product_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createCategory(CreateCategoryInput: CreateCategoryInput): Nullable<string> | Promise<Nullable<string>>;

    abstract deleteCategory(id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract likeProduct(variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract removeLikeProduct(like_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract addToCart(variation_id: string, quantity: number): Nullable<string> | Promise<Nullable<string>>;

    abstract removeFromCart(cart_item_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createOrder(): Nullable<string> | Promise<Nullable<string>>;
}

export type DateTime = any;
type Nullable<T> = T | null;
