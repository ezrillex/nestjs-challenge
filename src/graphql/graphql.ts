
/*
 * -------------------------------------------------------
 * THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
 * -------------------------------------------------------
 */

/* tslint:disable */
/* eslint-disable */

export class GetProductsInput {
    first?: Nullable<number>;
    offset?: Nullable<number>;
    categoryFilter?: Nullable<string[]>;
    search?: Nullable<string>;
    likedOnly?: Nullable<boolean>;
    omitImages?: Nullable<boolean>;
    sort?: Nullable<string>;
}

export class GetOrdersInput {
    first?: Nullable<number>;
    offset?: Nullable<number>;
    client_id?: Nullable<string>;
}

export class CreateProductInput {
    name: string;
    description: string;
    categories?: Nullable<string[]>;
    variations: CreateVariationInput[];
}

export class CreateVariationInput {
    title: string;
    price: number;
    stock: number;
}

export class UpdateProductInput {
    id: string;
    name?: Nullable<string>;
    description?: Nullable<string>;
    categories?: Nullable<string[]>;
    is_published?: Nullable<boolean>;
}

export class UpdateProductVariationInput {
    id: string;
    title?: Nullable<string>;
    price?: Nullable<number>;
    stock?: Nullable<number>;
}

export class CreateProductVariationInput {
    product_id: string;
    title: string;
    price: number;
    stock: number;
}

export class Categories {
    __typename?: 'Categories';
    id: string;
    name: string;
}

export class GetCategoriesResponse {
    __typename?: 'GetCategoriesResponse';
    id: string;
    name: string;
    products: Products[];
}

export class Images {
    __typename?: 'Images';
    id: string;
    url: string;
    created_at: DateTime;
    product_variation_id?: Nullable<string>;
}

export class ProductVariations {
    __typename?: 'ProductVariations';
    id: string;
    title: string;
    price: number;
    stock: number;
    images?: Nullable<Images[]>;
}

export class Products {
    __typename?: 'Products';
    id: string;
    is_published: boolean;
    is_deleted: boolean;
    name: string;
    description: string;
    variations: ProductVariations[];
    categories: Categories[];
    created_by: string;
    created_at: DateTime;
    last_updated_by?: Nullable<string>;
    last_updated_at?: Nullable<DateTime>;
}

export class LikesOfProducts {
    __typename?: 'LikesOfProducts';
    user_id: string;
    product_variation_id: string;
    state: boolean;
}

export class Users {
    __typename?: 'Users';
    id: string;
    first_name: string;
    last_name: string;
    email: string;
    role: string;
    created_at: DateTime;
    login_at: DateTime;
    password_last_updated: DateTime;
    likes_products: LikesOfProducts[];
}

export class CartItems {
    __typename?: 'CartItems';
    id: string;
    cart_owner: Users;
    product_variation: ProductVariations;
    quantity: number;
}

export class OrderItems {
    __typename?: 'OrderItems';
    id: string;
    product_variation: ProductVariations;
    quantity: number;
    price_purchased_at: number;
}

export class PaymentIntents {
    __typename?: 'PaymentIntents';
    id: string;
    status?: Nullable<string>;
    created_at: DateTime;
    stripe_event_id: string;
}

export class Orders {
    __typename?: 'Orders';
    id: string;
    user: Users;
    payments: PaymentIntents[];
    paymentStatus: string;
    orderStatus: string;
    order_items: OrderItems[];
}

export abstract class IQuery {
    __typename?: 'IQuery';

    abstract user(id: string): Users | Promise<Users>;

    abstract getLikes(): Nullable<ProductVariations[]> | Promise<Nullable<ProductVariations[]>>;

    abstract getProducts(GetProductsInput: GetProductsInput): Nullable<Products[]> | Promise<Nullable<Products[]>>;

    abstract getProductById(id: string): Nullable<Products> | Promise<Nullable<Products>>;

    abstract getCategories(): Nullable<GetCategoriesResponse[]> | Promise<Nullable<GetCategoriesResponse[]>>;

    abstract getCartItems(): Nullable<CartItems[]> | Promise<Nullable<CartItems[]>>;

    abstract getOrders(GetOrdersInput: GetOrdersInput): Nullable<Orders[]> | Promise<Nullable<Orders[]>>;

    abstract getOrder(order_id: string): Nullable<Orders> | Promise<Nullable<Orders>>;
}

export abstract class IMutation {
    __typename?: 'IMutation';

    abstract toggleLike(product_variation_id: string): Nullable<LikesOfProducts> | Promise<Nullable<LikesOfProducts>>;

    abstract createProduct(CreateProductInput: CreateProductInput): Nullable<Products> | Promise<Nullable<Products>>;

    abstract updateProduct(UpdateProductInput: UpdateProductInput): Nullable<Products> | Promise<Nullable<Products>>;

    abstract deleteProduct(product_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract updateProductVariation(UpdateProductVariationInput: UpdateProductVariationInput): Nullable<ProductVariations> | Promise<Nullable<ProductVariations>>;

    abstract createProductVariation(CreateProductVariationInput: CreateProductVariationInput): Nullable<ProductVariations> | Promise<Nullable<ProductVariations>>;

    abstract deleteProductVariation(variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createCategory(name: string): Nullable<Categories> | Promise<Nullable<Categories>>;

    abstract deleteCategory(id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract addToCart(variation_id: string, quantity: number): Nullable<CartItems> | Promise<Nullable<CartItems>>;

    abstract removeFromCart(product_variation_id: string): Nullable<string> | Promise<Nullable<string>>;

    abstract createOrder(): Nullable<Orders> | Promise<Nullable<Orders>>;
}

export type DateTime = any;
type Nullable<T> = T | null;
