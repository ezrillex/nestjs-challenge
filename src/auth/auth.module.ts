import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { AuthController } from './auth.controller';
import { AuthService } from './auth.service';
import { PrismaService } from '../prisma/prisma.service';
import { APP_GUARD } from '@nestjs/core';
import { AuthGuard } from './guards/auth.guard';
import { RolesGuard } from './guards/roles.guard';
import { ThrottlerModule } from '@nestjs/throttler';
import { EmailsModule } from '../emails/emails.module';

@Module({
  controllers: [AuthController],
  providers: [
    AuthService,
    PrismaService,
    {
      provide: APP_GUARD,
      useClass: AuthGuard,
    },
    {
      provide: APP_GUARD,
      useClass: RolesGuard,
    },
  ],
  imports: [
    JwtModule.registerAsync({
      imports: [ConfigModule],
      global: true,
      useFactory: async (configService: ConfigService) => ({
        secret: configService.get<string>('JWTCONSTANT'),
        signOptions: { expiresIn: '24h' },
      }),
      inject: [ConfigService],
    }),
    ThrottlerModule.forRoot([{ ttl: 60000, limit: 100 }]),
    EmailsModule,
  ],
})
export class AuthModule {}
