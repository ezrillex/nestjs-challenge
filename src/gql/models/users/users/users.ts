import { Field, ID, ObjectType } from '@nestjs/graphql';
import { LikesOfProducts } from '../../products/likes/likes_of_products/likes_of_products';
import { IsEmail } from 'class-validator';

// partial implementation of actual table due to graphql not dealing with auth and roles
@ObjectType()
export class Users {
  @Field(() => ID)
  id: string;

  @Field(() => String)
  first_name: string;

  @Field(() => String)
  last_name: string;

  @IsEmail()
  @Field(() => String)
  email: string;

  @Field(() => [LikesOfProducts])
  likes_products: LikesOfProducts[];
}
