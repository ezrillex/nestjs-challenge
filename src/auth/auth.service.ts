import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { SignupUserDto } from './dto/signup-user.dto';
import { PrismaService } from '../prisma/prisma.service';
import * as bcrypt from 'bcrypt';
import { ConfigService } from '@nestjs/config';
import { roles } from '@prisma/client';
import { DateTime } from 'luxon';
import { LoginUserDto } from './dto/login-user.dto';
import { JwtService } from '@nestjs/jwt';
import { ForgotPasswordDto } from './dto/forgot-password.dto';
import { ResetPasswordDto } from './dto/reset-password.dto';
import { EMAIL_TEMPLATE, EmailsService } from '../emails/emails.service';
import { UsersService } from '../users/users.service';
import { Users } from '../users/users.model';
import { ForgotPasswordResponseDto } from './dto/forgot-password-response.dto';
import { ResetPasswordResponseDto } from './dto/ResetPasswordResponse.dto';

// todo refactor all specific user related operations to user service.
@Injectable()
export class AuthService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly configService: ConfigService,
    private jwtService: JwtService,
    private readonly emailsService: EmailsService,
    private readonly usersService: UsersService,
  ) {}

  /*
   @description method throws error if timestamps are in 24-hour period.
   @param timestamps an array of timestamps
   @param message customize the error message
   @param status customize the http status code
   */
  checkIfAttemptsInLast24h(
    timestamps: Date[],
    message: string = 'Too many attempts in 24 hour period.',
    status: HttpStatus = HttpStatus.FORBIDDEN,
  ): void {
    // account locked check
    if (timestamps && timestamps.length >= 3) {
      const aDayAgo = DateTime.now().minus({ days: 1 });

      const isIn24h = timestamps.reduce((accumulator, current) => {
        const value = DateTime.fromISO(current.toISOString());
        accumulator = accumulator === value >= aDayAgo; // basically an AND of whole array to check if is past 24 hours.
        return accumulator;
      }, true);

      if (isIn24h) {
        throw new HttpException(message, status);
      }
    }
  }

  async register(data: SignupUserDto): Promise<Users> {
    if (data.password !== data.repeat_password) {
      throw new HttpException('Passwords do not match', HttpStatus.BAD_REQUEST);
    }

    const duplicate = await this.prisma.users.count({
      where: { email: data.email },
    });
    if (duplicate > 0) {
      throw new HttpException('Email is already in use', HttpStatus.CONFLICT);
    }

    const hashed = await bcrypt.hash(data.password, 10);
    let role: roles = roles.customer;
    if (this.configService.get<string>('AUTO_ROLE') === 'TRUE') {
      const email = data.email.split('@');
      const suffix = email[0].split('+');
      if (suffix.length > 1) {
        if (suffix.at(-1) === 'admin') {
          role = roles.admin;
        } else if (suffix.at(-1) === 'manager') {
          role = roles.manager;
        }
      }
    }

    const user = this.prisma.users.create({
      data: {
        first_name: data.first_name,
        last_name: data.last_name,
        email: data.email,
        password: hashed,
        role,
      },
      select: {
        id: true,
        email: true,
        role: true,
        last_name: true,
        first_name: true,
        created_at: true,
      },
    });

    await this.emailsService.sendEmail(EMAIL_TEMPLATE.WELCOME, {
      user_first_name: data.first_name,
    });

    return user;
  }

  async recordFailedLoginAttempt(
    id: string,
    timestamps: Date[],
  ): Promise<Users> {
    const ts_length = timestamps.push(new Date());
    if (ts_length > 3) {
      timestamps.shift();
    }

    return this.prisma.users.update({
      where: { id: id },
      data: {
        failed_login_attempts: timestamps.length,
        failed_login_attempts_timestamps: timestamps,
      },
    });
  }

  async recordSuccessfulLoginAttempt(
    id: string,
    token: string,
  ): Promise<Users> {
    const now = new Date().toISOString();
    return this.prisma.users.update({
      where: { id: id },
      data: { login_at: now, session_token: token },
    });
  }

  async recordPasswordResetRequest(
    id: string,
    timestamps: Date[],
    resetToken: string,
  ): Promise<Users> {
    const ts_length = timestamps.push(new Date());
    if (ts_length > 3) {
      timestamps.shift();
    }

    return this.prisma.users.update({
      where: { id: id },
      data: {
        password_reset_requests: timestamps.length,
        password_reset_requests_timestamps: timestamps,
        password_reset_token: resetToken,
      },
    });
  }

  async forgotPassword(
    data: ForgotPasswordDto,
  ): Promise<ForgotPasswordResponseDto> {
    const user = await this.usersService.getUserByEmail(data.email);

    // too many reset attempts check
    this.checkIfAttemptsInLast24h(
      user.password_reset_requests_timestamps,
      'Too many password reset requests in a 24 hour period. Try again later.',
    );

    const token = this.jwtService.sign({
      user: user.id,
    });

    const forgotPasswordData = await this.recordPasswordResetRequest(
      user.id,
      user.password_reset_requests_timestamps,
      token,
    );

    await this.emailsService.sendEmail(EMAIL_TEMPLATE.FORGOT_PASSWORD, {
      reset_token: token,
    });

    return {
      message:
        'An email has been sent with a link to reset the password. Check your email.',
      request_count: forgotPasswordData.password_reset_requests,
      request_timestamps: forgotPasswordData.password_reset_requests_timestamps,
    };
  }

  async resetPasswordWithToken(
    id: string,
    new_password: string,
  ): Promise<Users> {
    const hashed = await bcrypt.hash(new_password, 10);

    return this.prisma.users.update({
      where: { id: id },
      data: {
        password: hashed,
        password_last_updated: new Date(),
        password_reset_token: null,
        // clear previous attempts
        password_reset_requests: 0,
        password_reset_requests_timestamps: [],
        // unlocks account
        failed_login_attempts: 0,
        failed_login_attempts_timestamps: [],
      },
    });
  }

  async logout(id: string): Promise<Users> {
    const now = new Date().toISOString();
    return this.prisma.users.update({
      where: { id: id },
      data: {
        session_token: null,
        logout_at: now,
      },
    });
  }

  async login(data: LoginUserDto): Promise<Users & { token: string }> {
    const user = await this.usersService.getUserByEmail(data.email);

    // account locked check
    this.checkIfAttemptsInLast24h(
      user.failed_login_attempts_timestamps,
      'Too many failed login attempts, account is locked. Try again later.',
    );
    const isCorrectPassword = await bcrypt.compare(
      data.password,
      user.password,
    );
    if (isCorrectPassword) {
      const token = await this.jwtService.signAsync({
        user: user.id,
        role: user.role,
      });

      await this.recordSuccessfulLoginAttempt(user.id, token);
      await this.emailsService.sendEmail(EMAIL_TEMPLATE.LOGIN_SUCCESSFUL);
      return {
        id: user.id,
        token,
        first_name: user.first_name,
        last_name: user.last_name,
        email: user.email,
        created_at: user.created_at,
        role: user.role,
        login_at: user.login_at,
        password_last_updated: user.password_last_updated,
      };
    } else {
      await this.recordFailedLoginAttempt(
        user.id,
        user.failed_login_attempts_timestamps,
      );
      await this.emailsService.sendEmail(EMAIL_TEMPLATE.LOGIN_FAIL);
      throw new HttpException(
        'Wrong password. Try again or use the forgot password api to reset it.',
        HttpStatus.UNAUTHORIZED,
      );
    }
  }

  async changePassword(
    data: ResetPasswordDto,
  ): Promise<ResetPasswordResponseDto> {
    if (data.password !== data.repeat_password) {
      throw new HttpException('Passwords do not match', HttpStatus.BAD_REQUEST);
    }
    let payload: { user: string };
    try {
      payload = await this.jwtService.verifyAsync(data.reset_token);
    } catch (error) {
      throw new HttpException(
        `An error occurred when validating reset token: ${error.message}`,
        HttpStatus.UNAUTHORIZED,
      );
    }
    const { password_reset_token, id } = await this.usersService.getUserById(
      payload.user,
    );

    if (password_reset_token === data.reset_token) {
      const { password_last_updated } = await this.resetPasswordWithToken(
        id,
        data.password,
      );
      return {
        message: 'The password has been reset successfully!',
        password_last_updated,
      };
    } else {
      throw new HttpException('The token is not valid.', HttpStatus.FORBIDDEN);
    }
  }
}
