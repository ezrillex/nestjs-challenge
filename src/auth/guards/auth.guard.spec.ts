import { Test, TestingModule } from '@nestjs/testing';
import { AuthGuard } from './auth.guard';
import { Reflector } from '@nestjs/core';
import { JwtService } from '@nestjs/jwt';
import { AuthService } from '../auth.service';
import { PrismaService } from '../../prisma/prisma.service';
import { ConfigService } from '@nestjs/config';
import { ExecutionContext } from '@nestjs/common';
import { GqlExecutionContext } from '@nestjs/graphql';
import { Users } from '@prisma/client';
import { UsersService } from '../../users/users.service';
import { EmailsService } from '../../emails/emails.service';

jest.mock('@sendgrid/mail', () => {
  return {
    __esModule: true, // Indicates that this is a module with default export
    default: {
      setApiKey: jest.fn(), // Mock setApiKey method
      send: jest.fn(), // Mock send method
    },
  };
});

describe('Auth Guard', () => {
  let guard: AuthGuard;
  let reflector: Reflector;
  let authService: AuthService;
  let jwtService: JwtService;
  let usersService: UsersService;
  let emailsService: EmailsService;
  let context: ExecutionContext;
  let requestMock: { public_private_mode?: any; headers: any; user?: any };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        AuthGuard,
        PrismaService,
        ConfigService,
        AuthService,
        Reflector,
        JwtService,
        UsersService,
        EmailsService,
      ],
    }).compile();

    guard = module.get<AuthGuard>(AuthGuard);
    reflector = module.get<Reflector>(Reflector);
    authService = module.get<AuthService>(AuthService);
    jwtService = module.get<JwtService>(JwtService);
    usersService = module.get<UsersService>(UsersService);
    emailsService = module.get<EmailsService>(EmailsService);

    requestMock = {
      headers: {},
    };

    context = {
      getHandler: jest.fn(),
      getClass: jest.fn(),
      getType: jest.fn().mockReturnValue('http'),
      switchToHttp: jest.fn().mockReturnValue({
        getRequest: jest.fn().mockReturnValue(requestMock),
      }),
    } as unknown as ExecutionContext;
  });

  it('should be defined', () => {
    expect(guard).toBeDefined();
  });

  it('should implement can activate', () => {
    expect(guard.canActivate).toBeInstanceOf(Function);
  });

  it('should return true if endpoint is public', async () => {
    jest.spyOn(reflector, 'getAllAndOverride').mockReturnValue(true);
    await expect(guard.canActivate(context)).resolves.toEqual(true);
  });

  it('should return true if endpoint is public/private and no token provided must go into public mode', async () => {
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(guard.canActivate(context)).resolves.toEqual(true);
    expect(requestMock.public_private_mode).toEqual('public');
  });

  it('should throw if endpoint is public/private and no token is provided', async () => {
    requestMock.headers.authorization = 'Bearer';
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(
      guard.canActivate(context),
    ).rejects.toThrowErrorMatchingSnapshot('Token missing or invalid error');
    // restore for other tests
    delete requestMock.headers.authorization;
  });

  it('should throw if endpoint is public/private and an invalid token is provided', async () => {
    requestMock.headers.authorization = 'Bearer SPAIN-GO';
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(
      guard.canActivate(context),
    ).rejects.toThrowErrorMatchingSnapshot('Token invalid error');
    // restore for other tests
    delete requestMock.headers.authorization;
  });

  it('should throw if endpoint is public/private and a valid token is provided, but there are no active sessions', async () => {
    jest
      .spyOn(jwtService, 'verifyAsync')
      .mockResolvedValue({ user: 'some user id' });
    jest
      .spyOn(usersService, 'getUserById')
      .mockResolvedValue({ id: 'some user id' } as Users);
    requestMock.headers.authorization = 'Bearer SPAIN-GO';
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(
      guard.canActivate(context),
    ).rejects.toThrowErrorMatchingSnapshot('No active sessions error');
    // restore for other tests
    delete requestMock.headers.authorization;
  });

  it('should throw if endpoint is public/private and a valid token is provided, but token does not match current session token', async () => {
    jest
      .spyOn(jwtService, 'verifyAsync')
      .mockResolvedValue({ user: 'some user id' });
    jest.spyOn(usersService, 'getUserById').mockResolvedValue({
      id: 'some user id',
      session_token: 'JAPAN-GO',
    } as Users);
    requestMock.headers.authorization = 'Bearer SPAIN-GO';
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(
      guard.canActivate(context),
    ).rejects.toThrowErrorMatchingSnapshot('Not the last session token error');
    // restore for other tests
    delete requestMock.headers.authorization;
  });

  it('should return true if endpoint is public/private and a valid token is provided and mode is private and user is set', async () => {
    const user = {
      id: 'some user id',
      session_token: 'SPAIN-GO',
      login_at: new Date(),
    } as Users;
    jest
      .spyOn(jwtService, 'verifyAsync')
      .mockResolvedValue({ user: 'some user id' });
    jest.spyOn(usersService, 'getUserById').mockResolvedValue(user);
    requestMock.headers.authorization = 'Bearer SPAIN-GO';
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(guard.canActivate(context)).resolves.toEqual(true);
    expect(requestMock.user).toEqual(user);
    expect(requestMock.public_private_mode).toEqual('private');
    // restore for other tests
    delete requestMock.headers.authorization;
    delete requestMock.user;
    delete requestMock.public_private_mode;
  });
  it('should throw if endpoint is public/private and a valid token is provided but is older than 3 days', async () => {
    const user = {
      id: 'some user id',
      session_token: 'SPAIN-GO',
      login_at: new Date(2020, 12, 12, 12, 12, 12),
    } as Users;
    jest
      .spyOn(jwtService, 'verifyAsync')
      .mockResolvedValue({ user: 'some user id' });
    jest.spyOn(usersService, 'getUserById').mockResolvedValue(user);
    requestMock.headers.authorization = 'Bearer SPAIN-GO';
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(
      guard.canActivate(context),
    ).rejects.toThrowErrorMatchingSnapshot('token expired older 3 days');
    // restore for other tests
    delete requestMock.headers.authorization;
  });

  it('GRAPHQL should return true if endpoint is public/private and a valid token is provided and mode is private and user is set', async () => {
    const gqlRequest = {
      headers: {
        authorization: 'Bearer SPAIN-GO',
      },
    };

    context.getType = jest.fn().mockReturnValue('graphql');
    jest.spyOn(GqlExecutionContext, 'create').mockReturnValue({
      getContext: jest.fn().mockReturnValue({
        req: gqlRequest,
      }),
    } as unknown as GqlExecutionContext);

    const user = {
      id: 'some user id',
      session_token: 'SPAIN-GO',
      login_at: new Date(),
    } as Users;
    jest
      .spyOn(jwtService, 'verifyAsync')
      .mockResolvedValue({ user: 'some user id' });
    jest.spyOn(usersService, 'getUserById').mockResolvedValue(user);
    requestMock.headers.authorization = 'Bearer SPAIN-GO';
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(guard.canActivate(context)).resolves.toEqual(true);
    expect(gqlRequest['user']).toEqual(user);
    expect(gqlRequest['public_private_mode']).toEqual('private');
  });

  it('GRAPHQL throws if endpoint is public/private and an invalid token is provided', async () => {
    const gqlRequest = {
      headers: {
        authorization: 'Bearer JAPAN-GO',
      },
    };

    context.getType = jest.fn().mockReturnValue('graphql');
    jest.spyOn(GqlExecutionContext, 'create').mockReturnValue({
      getContext: jest.fn().mockReturnValue({
        req: gqlRequest,
      }),
    } as unknown as GqlExecutionContext);

    const user = {
      id: 'some user id',
      session_token: 'SPAIN-GO',
      login_at: new Date(),
    } as Users;
    jest
      .spyOn(jwtService, 'verifyAsync')
      .mockResolvedValue({ user: 'some user id' });
    jest.spyOn(usersService, 'getUserById').mockResolvedValue(user);
    requestMock.headers.authorization = 'Bearer SPAIN-GO';
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(
      guard.canActivate(context),
    ).rejects.toThrowErrorMatchingSnapshot('token invalid graphql');
  });

  it('GRAPHQL returns true if endpoint is public/private and no token is provided, mode is set to public', async () => {
    const gqlRequest = {
      headers: {},
    };

    context.getType = jest.fn().mockReturnValue('graphql');
    jest.spyOn(GqlExecutionContext, 'create').mockReturnValue({
      getContext: jest.fn().mockReturnValue({
        req: gqlRequest,
      }),
    } as unknown as GqlExecutionContext);

    const user = {
      id: 'some user id',
      session_token: 'SPAIN-GO',
      login_at: new Date(),
    } as Users;
    jest
      .spyOn(jwtService, 'verifyAsync')
      .mockResolvedValue({ user: 'some user id' });
    jest.spyOn(usersService, 'getUserById').mockResolvedValue(user);
    requestMock.headers.authorization = 'Bearer SPAIN-GO';
    jest
      .spyOn(reflector, 'getAllAndOverride')
      .mockReturnValueOnce(undefined) // is public
      .mockReturnValueOnce(true); // is public private
    await expect(guard.canActivate(context)).resolves.toEqual(true);
  });
});
