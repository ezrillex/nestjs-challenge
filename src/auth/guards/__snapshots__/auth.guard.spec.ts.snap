// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Guard GRAPHQL throws if endpoint is public/private and an invalid token is provided: token invalid graphql 1`] = `"The provided token is no longer valid. Make sure to authenticate with your most up-to-date token."`;

exports[`Auth Guard should throw if endpoint is public/private and a valid token is provided but is older than 3 days: token expired older 3 days 1`] = `"Token has expired."`;

exports[`Auth Guard should throw if endpoint is public/private and a valid token is provided, but there are no active sessions: No active sessions error 1`] = `"User has no active sessions."`;

exports[`Auth Guard should throw if endpoint is public/private and a valid token is provided, but token does not match current session token: Not the last session token error 1`] = `"The provided token is no longer valid. Make sure to authenticate with your most up-to-date token."`;

exports[`Auth Guard should throw if endpoint is public/private and an invalid token is provided: Token invalid error 1`] = `"Bearer token is invalid."`;

exports[`Auth Guard should throw if endpoint is public/private and no token is provided: Token missing or invalid error 1`] = `"Bearer token is missing or is invalid."`;
